db/db.go

```diff
@@ -83,149 +64,256 @@ type Database struct {
 		if err != nil {
-			db.metadataError = errors.Annotate(err,
+			r.metadataError = errors.Annotate(err,
 				"failed to open file for reading: '%s'", fileName)
 		}
 		defer f.Close()
 
 		dec := json.NewDecoder(f)
-		if err := dec.Decode(&db.metadata); err != nil {
-			db.metadataError = errors.Annotate(err, "failed to decode JSON")
+		if err := dec.Decode(&r.metadata); err != nil {
+			r.metadataError = errors.Annotate(err, "failed to decode JSON")
 		}
 	})
-	return db.metadataError
+	return r.metadataError
 }
 
-func (db *Database) cacheTickers() error {
-	db.tickersOnce.Do(func() {
-		if err := readGob(db.tickersFile(), &db.tickers); err != nil {
-			db.tickersError = errors.Annotate(
-				err, "failed to load %s", db.tickersFile())
+func (r *Reader) cacheTickers() error {
+	r.tickersOnce.Do(func() {
+		if err := readGob(tickersFile(r.cachePath), &r.tickers); err != nil {
+			r.tickersError = errors.Annotate(
+				err, "failed to load %s", tickersFile(r.cachePath))
 		}
 	})
-	return db.tickersError
+	return r.tickersError
 }
 
-func (db *Database) cacheActions() error {
-	db.actionsOnce.Do(func() {
-		if err := readGob(db.actionsFile(), &db.actions); err != nil {
-			db.actionsError = errors.Annotate(
-				err, "failed to load %s", db.actionsFile())
+func (r *Reader) cacheActions() error {
+	r.actionsOnce.Do(func() {
+		if err := readGob(actionsFile(r.cachePath), &r.actions); err != nil {
+			r.actionsError = errors.Annotate(
+				err, "failed to load %s", actionsFile(r.cachePath))
 		}
 	})
-	return db.actionsError
+	return r.actionsError
 }
 
-func (db *Database) cacheMonthly() error {
-	db.monthlyOnce.Do(func() {
-		if err := readGob(db.monthlyFile(), &db.monthly); err != nil {
-			db.monthlyError = errors.Annotate(
-				err, "failed to load %s", db.monthlyFile())
+func (r *Reader) cacheMonthly() error {
+	r.monthlyOnce.Do(func() {
+		if err := readGob(monthlyFile(r.cachePath), &r.monthly); err != nil {
+			r.monthlyError = errors.Annotate(
+				err, "failed to load %s", monthlyFile(r.cachePath))
 		}
 	})
-	return db.monthlyError
+	return r.monthlyError
+}
+
+// Metadata for the database. It is cached in memory upon the first call.
+func (r *Reader) Metadata() (Metadata, error) {
+	if err := r.cacheMetadata(); err != nil {
+		return Metadata{}, errors.Annotate(err, "failed to load metadata")
+	}
+	return r.metadata, nil
+}
+
+// TickerRow for the given ticker. It's an error if a ticker is not in R.
+// Tickers are cached in memory upon the first call. Go routine safe.
+func (r *Reader) TickerRow(ticker string) (TickerRow, error) {
+	if err := r.cacheTickers(); err != nil {
+		return TickerRow{}, errors.Annotate(err, "failed to load tickers")
+	}
+	row, ok := r.tickers[ticker]
+	if !ok {
+		return TickerRow{}, errors.Reason("no such ticker: %s", ticker)
+	}
+	return row, nil
+}
+
+// Tickers returns the list of tickers satisfying Reader's constraints.
+// Tickers are cached in memory upon the first call. Go routine safe.
```

Comment:
All tickers are cached in memory. Modifying constraints takes effect for the next call. Go routine safe assuming constraints are not modified.

Reply:
Updated here and below.